python/pyspark/ml/image.py

```diff
@@ -0,0 +1,133 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import pyspark
+from pyspark import SparkContext
+from pyspark.sql.types import *
+from pyspark.sql.types import Row, _create_row
+from pyspark.sql import DataFrame
+from pyspark.ml.param.shared import *
+import numpy as np
+
+undefinedImageType = "Undefined"
+
+ImageFields = ["origin", "height", "width", "nChannels", "mode", "data"]
+
+ocvTypes = {
+    undefinedImageType: -1,
+    "CV_8U": 0, "CV_8UC1": 0, "CV_8UC2": 8, "CV_8UC3": 16, "CV_8UC4": 24,
+    "CV_8S": 1, "CV_8SC1": 1, "CV_8SC2": 9, "CV_8SC3": 17, "CV_8SC4": 25,
+    "CV_16U": 2, "CV_16UC1": 2, "CV_16UC2": 10, "CV_16UC3": 18, "CV_16UC4": 26,
+    "CV_16S": 3, "CV_16SC1": 3, "CV_16SC2": 11, "CV_16SC3": 19, "CV_16SC4": 27,
+    "CV_32S": 4, "CV_32SC1": 4, "CV_32SC2": 12, "CV_32SC3": 20, "CV_32SC4": 28,
+    "CV_32F": 5, "CV_32FC1": 5, "CV_32FC2": 13, "CV_32FC3": 21, "CV_32FC4": 29,
+    "CV_64F": 6, "CV_64FC1": 6, "CV_64FC2": 14, "CV_64FC3": 22, "CV_64FC4": 30
+}
+
+ImageSchema = StructType([
+    StructField(ImageFields[0], StringType(),  True),
+    StructField(ImageFields[1], IntegerType(), False),
+    StructField(ImageFields[2], IntegerType(), False),
+    StructField(ImageFields[3], IntegerType(), False),
+    # OpenCV-compatible type: CV_8UC3 in most cases
+    StructField(ImageFields[4], StringType(), False),
+    # bytes in OpenCV-compatible order: row-wise BGR in most cases
+    StructField(ImageFields[5], BinaryType(), False)])
+
+
+# TODO: generalize to other datatypes and number of channels
+def toNDArray(image):
+    """
+    Converts an image to a 1-dimensional array
+
+    Args:
+        image (object): The image to be converted
+
+    Returns:
+        array: The image as a 1-dimensional array
+
+    .. versionadded:: 2.3.0
+    """
+    height = image.height
+    width = image.width
+    return np.asarray(image.data, dtype=np.uint8) \
+             .reshape((height, width, 3))[:, :, (2, 1, 0)]
```

Comment:
This code assumes `image` is 3-channel GBR image and the user wants an RGB image. I think we should try and support at least the 3 image types that `readImages` (CV_8UC1, CV_8UC3, and CV_8UC4) but it would be nice to also support 1, 3, and 4 channel float images.

The ndarray constructor is quite flexible and might be easier to work with than calling `asarray` in this case because you want to treat the bytearray as a buffer not as a sequence of ints.

```
np.ndarray(
  shape=(height, width, nChannels),
  dtype=np.uint8,
  buffer=image.data,
  strides=(width * nChannels, nChannels, 1))
```

Also I have mixed feelings about re-ordering the channels. I think it's probably useful in the most common use-case, but (if my understanding is correct) the open cv types don't require a specific channel order so we can't just assume the input is RGB or RGBA. Maybe we should avoid re-ordering, document the ordering we use whenever appropriate, and then leave it up to the user to do any necessary re-order for themselves.

Reply:
Totally agree with the first suggestion -- we should cover 1,3 and 4-channel images here. I believe we have channel reordering to match OpenCV behavior for python; let me do a couple of experiments to confirm.

Reply:
I believe that OpenCV in python reads images using a BGR channel order by default. This conflicts with the RGB default used by some other python libraries (eg matplotlib) but since there doesn't seem to be a clear standard, I suggest we don't re-order anything by default. That way spark will at least maintain consistent behaviour between scala and python.

http://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_core/py_basic_ops/py_basic_ops.html

Reply:
Sorry for the late reply -- yes, you are totally right. OpenCV returns BGR, and we should do the same.